iceoryx_posh/include/iceoryx_posh/internal/gateway/gateway_generic.inl

```diff
@@ -37,9 +37,9 @@ inline GatewayGeneric<channel_t, gateway_t>::~GatewayGeneric() noexcept
 template <typename channel_t, typename gateway_t>
 inline void GatewayGeneric<channel_t, gateway_t>::runMultithreaded() noexcept
 {
+    m_isRunning.store(true, std::memory_order_relaxed);
```

Comment:
I think this is still bugged. As far as I understand the as-if rule https://en.cppreference.com/w/cpp/language/as_if it is still legal for the compiler to reorder this code to:
```cxx
    m_discoveryThread = std::thread([this] { this->discoveryLoop(); });
    m_forwardingThread = std::thread([this] { this->forwardingLoop(); });
    m_isRunning.store(true, std::memory_order_relaxed);
```

since this does not violate the as-if rule.

What we actually have to implement is much more complex. Either we use the memory order sequentially consistent for the load and store for `m_isRunning` or we introduce an atomic barrier which prevents an reordering based on the as-if rule.

We need to add the sequentially consistent for load and store since I think that one requires always a corresponding counterpart with the correct memory order otherwise one has again undefined behavior but I am unsure if this is also required for sequentially consistent - maybe not.

Reply:
AFAIK, reordering across thread starts is not allowed

https://en.cppreference.com/w/cpp/thread/thread/thread
https://stackoverflow.com/questions/73001485/c-compiler-reordering-read-and-write-across-thread-start

Reply:
I see. Well, I am actually wondering whether the use of the `std::memory_order_relaxed` memory order for the `m_isRunning` field is really needed at all. Obviously, this is not a very hot code, and using the default `std::memory_order_seq_cst` might be much easier to reason about, while losing completely negligible performance (if at all). So, I would propose to just use the sequentially consistent memory order and avoiding any premature optimization pitfalls.

Or am I missing some detail?

Reply:
@JakubSosnovec I'm fine either way. Using `std::memory_order_relaxed` should not be an issue and as you mentioned it is not a hot path so `std::memory_order_seq_cst` does not hurt. If it it changed, it should also be changed in `shutdown`, tough. The `while` loops should keep the current memory ordering.

Reply:
Alright, I updated it now.